src/Faker/Provider/nl_NL/Person.php

```diff
@@ -346,4 +379,75 @@ public function idNumber()
         }
         return implode('', array_reverse($nr));
     }
+
+    /**
+     * @param string|null $documentCode Key from $idStartLetter
+     * @param array $date               'timestamp' of 'type' issue or expiry. 'birth' is birthdate of person
+     * @param string|null $timezone     Timezone to use
+     */
+    public function identityDocument(
+        $documentCode = null,
+        $date = [
+            'timestamp'  => null,
+            'type'       => 'issue',
+            'birth'      => null,
+        ],
+        $timezone = null
+    ) {
+        // Validity
+        if (isset($date['birth'])) {
+            $age = $date['birth']->diff(new \DateTime('today'))->y;
+        }
+
+        if (isset($age) && $age < 18) {
+            $validity = 5;
+        } else {
+            $validity = 10;
+        }
+
+        // Date info
+        if (isset($date['type']) && $date['type'] === 'expiry') {
+            $dateType = 'expiry';
+        } else {
+            $dateType = 'issue';
+        }
+
+        $timeStamp = !isset($date['timestamp']) ? DateTime::dateTimeThisDecade('now', $timezone) : $date['timestamp'];
+        $ts = clone $timeStamp;
+        if ($dateType === 'expiry') {
+            $issueDate  = $ts->modify('-' . $validity . 'years');
+            $expiryDate = $timeStamp;
+        } else {
+            $issueDate  = $timeStamp;
+            $expiryDate = $ts->modify('+' . $validity . 'years');
+        }
+
+        // Document number
+        if (!isset($documentCode)) {
+            $documentCode = static::randomKey(static::$idStartLetter);
+        }
+        $startLetter = static::$idStartLetter[$documentCode];
+
+        // No '0' in documents issued from 1 december 2019
+        if ($issueDate < new \DateTime('2020-12-01')) {
+            $extra = [0];
+        }
+
+        $number = implode(
+            '',
+            array_merge(
+                [$startLetter],
+                [static::randomElement(static::$idLetters)],
+                static::randomElements(array_merge(static::$idLetters, static::$idNumbers, $extra), 7, true),
+                [static::randomElement(static::$idNumbers)],
+            )
+        );
+
+        return (object) [
```

Comment:
Not a fan of returning a casted object. Just return the is number that is generated as a string?

Most of the info here is put into the method als param

Reply:
By just returning the `number` you'd lose info about e.g. `issueDate` and `expiryDate` That info is calculated and not (all) passed as params.

Reply:
Because of the reasons from the previous comment, I would suggest returning only the number as string and add a separate method for `documentCode`. IssueDate can be simply passed as an argument, and expiryDate might be calculated by the client app.
From what I understand, BSN should match mod-11 checksum too: https://nl.wikipedia.org/wiki/Burgerservicenummer#11-proef